### PR TITLE
Reduce symbol array allocations during CSharpSemanticModel.LookupSymbolsInternal

### PR DESCRIPTION
**Pull request by ToddGrun**
During completion, I see this method generating results with several thousand entries which is evident in allocations in a profile I'm inspecting. This PR attempts to reduce intermediary arrays constructed during that process by modifying FilterNotReferenceable and FilterOverriddenOrHiddenIndexers to not require realization of an ImmutableArray as input.

Original allocations:
image

After PR allocations:
image

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 1]
```diff
@@ -1795,25 +1802,25 @@ private Symbol RemapSymbolIfNecessary(Symbol symbol)
...
             {
-                if (result.CanBeReferencedByName)
-                {
-                    builder?.Add(result);
-                }
-                else if (builder == null)
+                return sealedResults.AsImmutable();
+            }
+
+            var builder = ImmutableArray.CreateBuilder<ISymbol>(sealedResults.Count - cannotBeReferencedByNameCount);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1814]: Why wouldn't we want to use `ArrayBuilder` here? Either the size is compatible with the pool, and we'll avoid a new allocation, or we will get a new builder that will then get discarded?

File: src/Dependencies/Collections/SegmentedList`1.cs [hunk 2]
```diff
@@ -1161,6 +1162,17 @@ public bool TrueForAll(Predicate<T> match)
             return true;
         }
 
+        public ImmutableArray<T> ToImmutable()
```
Comment (333fred) [inline src/Dependencies/Collections/SegmentedList`1.cs:1165]: @sharwell has specific instructions for making changes to SegmentedList`1. I'm assuming that this change needs to be in a separate commit.

Comment (333fred) [reply src/Dependencies/Collections/SegmentedList`1.cs:1165]: Also, if we're going make this specific method, as opposed to just calling `ToImmutable` from the BCL or `SelectAsArray` with an identity selector, let's take advantage of array segments and copying entire chunks at a time, rather than just a `foreach` loop.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 3]
```diff
@@ -44,6 +45,8 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </remarks>
     internal abstract class CSharpSemanticModel : SemanticModel
     {
+        private static readonly ObjectPool<SegmentedList<ISymbol>> s_symbolListPool = new(() => new());
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:48]: We do not use `new()` in the compiler codebase.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 4]
```diff
@@ -1670,21 +1673,25 @@ private void CheckModelAndSyntaxNodeToSpeculate(CSharpSyntaxNode syntax)
...
 
-            ImmutableArray<ISymbol> sealedResults = results.ToImmutableAndFree();
-            return name == null
-                ? FilterNotReferencable(sealedResults)
-                : sealedResults;
+            var result = name == null
+                ? FilterNotReferenceable(builder)
+                : builder.ToImmutable();
+
+            builder.Clear();
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1688]: Let's supply a policy to the pool that clears on return so we don't have to remember all the time.

**Review by 333fred — COMMENTED** [review]
Done review pass.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1814]: There are three possibilities the way I see it:

1) ArrayBuilder finds an array that is too small and internally resizes it's builder to the new capacity. This allocates on retrieval but not on ToImmutableAndFree.
2) ArrayBuilder finds an array exactly the right size. This doesn't allocate upon retrieval or on ToImmutableAndFree.
3) ArrayBuilder finds an array that exceeds the requested size. This doesn't allocate upon retrieval but does on ToImmutableAndFree.

The only case where we allocate less is when the retrieved array is exactly the right size, but then we lost that entry from the pool. And all three cases require the ArrayBuilder to search it's pool for an available item.

I can switch to that if you prefer, but I don't believe it provides any performance benefit while needing to do the (slight) overhead of searching the pool for an available item.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1814]: Hmm. I think I misunderstood how ArrayBuilder's GetInstance works then. Let's leave a comment as to the reason that we aren't using it here.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 4 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1688]: What I really wanted was a PooledSegmentedList class similar to PooledHashSet/PooledDictionary. I wanted that so I wouldn't have to define the pool myself and that the pool could be shared with other code. I'm sure there is a good reason why Roslyn doesn't have that and requires everyone to create their pool, perhaps Sam knows?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 5]
```diff
@@ -4110,19 +4118,19 @@ private ImmutableArray<IPropertySymbol> GetIndexerGroupSemanticSymbols(BoundExpr
             }
 
             Binder binder = binderOpt ?? GetEnclosingBinder(GetAdjustedNodePosition(boundNode.Syntax));
-            var symbols = ArrayBuilder<ISymbol>.GetInstance();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: >var symbols = ArrayBuilder<ISymbol>.GetInstance(); [](http://example.com/codeflow?start=12&length=50)

Do we have an evidence that this code path has the same problem as ```LookupSymbolsInternal```? I can see how ```LookupSymbolsInternal``` could have thousands of items in the result when ```name``` is not specified. In this case, however, it looks like we are dealing with indexers in a specific type, there aren't usually that many of them.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 6]
```diff
@@ -4523,7 +4531,7 @@ private SymbolInfo GetNamedArgumentSymbolInfo(IdentifierNameSyntax identifierNam
             }
             else
             {
-                var symbols = ArrayBuilder<ISymbol>.GetInstance();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4526]: >var symbols = ArrayBuilder<ISymbol>.GetInstance(); [](http://example.com/codeflow?start=16&length=50)

Do we have an evidence that this code path has the same problem as ```LookupSymbolsInternal```? I can see how ```LookupSymbolsInternal``` could have thousands of items in the result when ```name``` is not specified. In this case, however, it looks like we are dealing with candidate symbols for a specific invocation/property access, there aren't usually that many of them.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 5 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: No, I have not seen this method in the profiles I was looking at. I was trying to be consistent with the other change away from ArrayBuilder<ISymbol> in this file, but I can change back to the original here if you prefer.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 6 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4526]: Same as before, this method didn't show up in the profile, but was trying to maintain consistency. Very willing to change back if preferred.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: >  I was trying to be consistent with the other change away from ArrayBuilder<ISymbol> in this file

I do not think we should be making changes based solely on this motive

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4526]: > Same as before, this method didn't show up in the profile, but was trying to maintain consistency. 

Same response here as well

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Reverted this one

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 5 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: Not reverting this one (yet) as it's a bit more entwined with the changes needed for LookupSymbolsInternal.

This method calls AppendSymbolsWithNameAndArity with the symbol list. 

LookupSymbolsInternal calls AppendSymbolsWithName which calls AppendSymbolsWithNameAndArity with the symbol list.

Changing one of these to SegmentedList but not the other would require duplicating AppendSymbolsWithNameAndArity to be implemented against both SegmentedList and ArrayBuilder (or just against a common base, but ArrayBuilder only implements IReadOnly interfaces).

Is that good enough justification to keep this part of the change, or would you prefer for it to be changed back to the original at the tradeoff of needing to (likely) duplicate at least some part of AppendSymbolsWithNameAndArity?

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: Maybe iteration 5 is preferable? It didn't require duplicating code and continues to use ArrayBuilder at this location.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4113]: > Maybe iteration 5 is preferable? It didn't require duplicating code and continues to use ArrayBuilder at this location.

If we are sharing a helper with ```LookupSymbolsInternal``` (sorry I didn't realize that before) that is good enough reason to use the same data structure here as well. I would go back to the iteration 3 version of this method and would get rid of the delegate.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7)

File: src/Dependencies/Collections/SegmentedList`1.cs (hunk 2 above)
Comment (ToddGrun) [reply src/Dependencies/Collections/SegmentedList`1.cs:1165]: @sharwell -- Should I move the SegmentedList.ToImmutable method to not be directly on SegmentedList?

Comment (sharwell) [reply src/Dependencies/Collections/SegmentedList`1.cs:1165]: I have two problems with this:

1. This method needs to be named `ToImmutableArray()` to avoid confusion with `ImmutableSegmentedList<T>`
2. This method should be an extension method instead of a member method on the type

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 7]
```diff
@@ -44,6 +45,8 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </remarks>
     internal abstract class CSharpSemanticModel : SemanticModel
     {
+        private static readonly ObjectPool<SegmentedList<ISymbol>> s_symbolListPool = new ObjectPool<SegmentedList<ISymbol>>(() => new SegmentedList<ISymbol>());
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:48]: For me, the gains claimed here aren't worth the complexity and risk brought in by another custom object pool. I believe this change can be reasonably reduced to:

1. Make `FilterNotReferencable` operate on `ArrayBuilder<T>` instead of `ImmutableArray<T>`
2. Make `FilterOverriddenOrHiddenIndexers` operate on `ArrayBuilder<T>` instead of `ImmutableArray<T>`

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:48]: Yeah, I'll switch this to not have the custom pool as it does complicate this change, and really only benefitted the cases where there were a large number of symbols.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 8]
```diff
@@ -1678,10 +1678,13 @@ private void CheckModelAndSyntaxNodeToSpeculate(CSharpSyntaxNode syntax)
...
 
-            ImmutableArray<ISymbol> sealedResults = results.ToImmutableAndFree();
-            return name == null
-                ? FilterNotReferencable(sealedResults)
-                : sealedResults;
+            var result = name == null
+                ? FilterNotReferenceable(results)
+                : results.ToImmutable();
+
+            results.Free();
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1685]: This can be improved by filtering in place:

```cs
if (name == null)
    FilterNotReferenceable(results);

var result = results.ToImmutableAndFree();
```

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:1685]: Let me give that a whirl, I think it's going to complicate FilterNotReferenceable a bit, but hopefully not too much

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Yep, I like that better.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 9]
```diff
@@ -4119,10 +4116,13 @@ private ImmutableArray<IPropertySymbol> GetIndexerGroupSemanticSymbols(BoundExpr
...
 
-            return FilterOverriddenOrHiddenIndexers(symbols.ToImmutableAndFree());
+            var result = FilterOverriddenOrHiddenIndexers(symbols);
+            symbols.Free();
+
+            return result;
         }
 
-        private static ImmutableArray<IPropertySymbol> FilterOverriddenOrHiddenIndexers(ImmutableArray<ISymbol> symbols)
+        private static ImmutableArray<IPropertySymbol> FilterOverriddenOrHiddenIndexers(ArrayBuilder<ISymbol> symbols)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4125]: >FilterOverriddenOrHiddenIndexers [](http://example.com/codeflow?start=55&length=32)

It looks like there is only one call site for this helper. Therefore, we could make it return void and simply remove filtered out items from the builder, if any.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 11). It looks like there an opportunity to avoid another array builder allocation.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 9 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4125]: Input array is of ISymbol, while output array is of type IPropertySymbol.

**Comment by ToddGrun** [comment]
@333fred for 2nd review

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler for a 2nd review

**Comment by ToddGrun** [comment]
Nudge to @dotnet/roslyn-compiler again for 2nd review

**Review by 333fred — APPROVED** [review]
